lib/src/kvs/ds.rs

```diff
@@ -821,34 +848,98 @@ impl Datastore {
 	// It is handy for testing, because it allows you to specify the timestamp,
 	// without depending on a system clock.
 	pub async fn tick_at(&self, ts: u64) -> Result<(), Error> {
-		self.save_timestamp_for_versionstamp(ts).await?;
+		let vs = self.save_timestamp_for_versionstamp(ts).await?;
+		if FFLAGS.change_feed_live_queries.enabled() {
+			self.process_lq_notifications(ts, vs).await?;
+		}
 		self.garbage_collect_stale_change_feeds(ts).await?;
 		// TODO Add LQ GC
 		// TODO Add Node GC?
 		Ok(())
 	}
 
 	// save_timestamp_for_versionstamp saves the current timestamp for the each database's current versionstamp.
-	pub(crate) async fn save_timestamp_for_versionstamp(&self, ts: u64) -> Result<(), Error> {
+	pub(crate) async fn save_timestamp_for_versionstamp(
+		&self,
+		ts: u64,
+	) -> Result<Option<Versionstamp>, Error> {
 		let mut tx = self.transaction(Write, Optimistic).await?;
-		if let Err(e) = self.save_timestamp_for_versionstamp_impl(ts, &mut tx).await {
-			return match tx.cancel().await {
-				Ok(_) => {
-					Err(e)
-				}
-				Err(txe) => {
-					Err(Error::Tx(format!("Error saving timestamp for versionstamp: {:?} and error cancelling transaction: {:?}", e, txe)))
+		match self.save_timestamp_for_versionstamp_impl(ts, &mut tx).await {
+			Ok(vs) => Ok(vs),
+			Err(e) => {
+				match tx.cancel().await {
+					Ok(_) => {
+						Err(e)
+					}
+					Err(txe) => {
+						Err(Error::Tx(format!("Error saving timestamp for versionstamp: {:?} and error cancelling transaction: {:?}", e, txe)))
+					}
 				}
+			}
+		}
+	}
+
+	pub async fn process_lq_notifications(
+		&self,
+		ts: u64,
+		vs: Option<Versionstamp>,
+	) -> Result<(), Error> {
+		// Return if there are no live queries
+		if self.notification_channel.is_none() {
+			return Ok(());
+		}
+		if self.local_live_queries.read().await.is_empty() {
+			return Ok(());
+		}
+
+		// Find live queries that need to catch up
+		let lock_local_lqs = self.local_live_queries.read().await;
+		let mut lqs_to_update: Vec<(LqIndexKey, LqIndexValue)> =
+			Vec::with_capacity(lock_local_lqs.len());
+		for (k, v) in lock_local_lqs.iter() {
+			// Skip if no need to update
+			if v.ts.value >= ts {
+				// This shouldn't be possible and/or may mean clock skew
+				continue;
 			};
+			if vs.is_some() && v.vs >= vs.unwrap() {
+				// This means that there hasn't been a change
+				continue;
+			};
+			// We know we need to process events, so we do so now
+			lqs_to_update.push((k.clone(), v.clone()))
+		}
+		if !lqs_to_update.is_empty() {
+			let vs = vs.expect(
+				"Can only track progress on live queries with a versionstamp from a change feed",
```

Comment:
Do we expect all live queries to have a `versionstamp`?

Reply:
Basically yes. You cannot have a live query without a change feed => having a change feed guarantees a versionstamp (I believe).